libraries/mediaviewer/api/src/main/kotlin/io/element/android/libraries/mediaviewer/api/local/pdf/PdfRendererManager.kt

```diff
@@ -25,20 +28,28 @@ class PdfRendererManager(
 ) {
     private val mutex = Mutex()
     private var pdfRenderer: PdfRenderer? = null
-    private val mutablePdfPages = MutableStateFlow<List<PdfPage>>(emptyList())
-    val pdfPages: StateFlow<List<PdfPage>> = mutablePdfPages
+    private val mutablePdfPages = MutableStateFlow<AsyncData<ImmutableList<PdfPage>>>(AsyncData.Uninitialized)
+    val pdfPages: StateFlow<AsyncData<ImmutableList<PdfPage>>> = mutablePdfPages
 
     fun open() {
         coroutineScope.launch {
             mutex.withLock {
                 withContext(Dispatchers.IO) {
-                    pdfRenderer = PdfRenderer(parcelFileDescriptor).apply {
-                        // Preload just 3 pages so we can render faster
-                        val firstPages = loadPages(from = 0, to = 3)
-                        mutablePdfPages.value = firstPages
-                        val nextPages = loadPages(from = 3, to = pageCount)
-                        mutablePdfPages.value = firstPages + nextPages
-                    }
+                    runCatching {
+                        PdfRenderer(parcelFileDescriptor)
+                    }.fold(
+                        onSuccess = { pdfRenderer ->
+                            this@PdfRendererManager.pdfRenderer = pdfRenderer
```

Comment:
Maybe directly do:

```kotlin
runCatching {
    pdfRenderer = PdfRenderer(parcelFileDescriptor)
}
```

So we get rid of the ugly `this@pdfRendererManager.pdfRenderer = ... ` line.

Reply:
ugly yes, but I need the `runCatching` to return the value.

Maybe using a also block like

```kotlin
runCatching {
    PdfRenderer(parcelFileDescriptor).also {
        pdfRenderer = it
    }
}.fold(
    onSuccess = { pdfRenderer ->
        // Preload just 3 pages so we can render faster
        val firstPages = pdfRenderer.loadPages(from = 0, to = 3)
        mutablePdfPages.value = AsyncData.Success(firstPages.toImmutableList())
        val nextPages = pdfRenderer.loadPages(from = 3, to = pdfRenderer.pageCount)
        mutablePdfPages.value = AsyncData.Success((firstPages + nextPages).toImmutableList())
    },
    onFailure = {
        mutablePdfPages.value = AsyncData.Failure(it)
    }
)
```

is less ugly?

Reply:
I have found another solution 55704f02273778214dfe8d184647b12e8d8440de

Reply:
I think the change is ok and maybe I'm overlooking something but with the change I proposed wouldn't the `pdfRenderer` property only be assigned if the initiailzation using the `parcelFileDescriptor` works too?

Reply:
Yes, but with the change you proposed, the result of the assignation is Unit, so I can't get the `PdfRenderer` instance in the `onSuccess` parameter block.